statecheck/expect_value_exists.go

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package statecheck
+
+import (
+	"context"
+	"fmt"
+
+	tfjson "github.com/hashicorp/terraform-json"
+
+	"github.com/hashicorp/terraform-plugin-testing/tfjsonpath"
+)
+
+// Resource State Check
+var _ StateCheck = expectValueExists{}
+
+type expectValueExists struct {
+	resourceAddress string
+	attributePath   tfjsonpath.Path
+}
+
+// CheckState implements the state check logic.
+func (e expectValueExists) CheckState(ctx context.Context, req CheckStateRequest, resp *CheckStateResponse) {
+	var resource *tfjson.StateResource
+
+	if req.State == nil {
+		resp.Error = fmt.Errorf("state is nil")
+
+		return
+	}
+
+	if req.State.Values == nil {
+		resp.Error = fmt.Errorf("state does not contain any state values")
+
+		return
+	}
+
+	if req.State.Values.RootModule == nil {
+		resp.Error = fmt.Errorf("state does not contain a root module")
+
+		return
+	}
+
+	for _, r := range req.State.Values.RootModule.Resources {
+		if e.resourceAddress == r.Address {
+			resource = r
+
+			break
+		}
+	}
+
+	if resource == nil {
+		resp.Error = fmt.Errorf("%s - Resource not found in state", e.resourceAddress)
+
+		return
+	}
+
+	_, err := tfjsonpath.Traverse(resource.AttributeValues, e.attributePath)
+
+	if err != nil {
+		resp.Error = err
+
+		return
+	}
+}
+
+// ExpectValueExists returns a state check that asserts that the specified
+// attribute at the given resource exists.
+func ExpectValueExists(resourceAddress string, attributePath tfjsonpath.Path) StateCheck {
+	return expectValueExists{
+		resourceAddress: resourceAddress,
+		attributePath:   attributePath,
+	}
+}
```

Comment:
Open question: Would this be better served with a "not null"/"any value" known value check? I think we can avoid this as its own top-level check so developers don't need to choose their preferred style.

Reply:
I think it's likely we'll all agree on using "not null"/"any value" known value check, so I'm implemented that and removed the top-level `ExpectValueExists` check. Happy to refactor/revert if we change our minds.